_7zz: correct license

06kellyjac opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

- `_7zz`: correct license and remove p7zip dependency
  - The code under `Compress/Rar*` is licensed under a specific unRAR license
  - Also `Compress/LzfseDecoder.cpp` is covered by BSD3
  - Meta has been updated and tweaked
  - Source is now downloaded from sourceforge in the `.tar.xz` version to avoid depending on `p7zip`
- `_7zz`: add notice of the license updates and optional unRAR licenced code


Resolves #163680


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [X] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [X] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

06kellyjac commented on pkgs/tools/archivers/7zz/default.nix:
```diff
@@ -14,17 +24,25 @@ stdenv.mkDerivation rec {
-  nativeBuildInputs = [ p7zip ] ++ lib.optionals useUasm [ uasm ];
+  nativeBuildInputs = lib.optionals useUasm [ uasm ];
+
+  # in addition to the make var, also remove Compress/Rar* code to be safe
+  postPatch = lib.optionalString (!enableUnfree) ''
+    rm -rf ../../Compress/Rar*
+  '';
```
this might be overkill but it shouldn't hurt :shrug:

06kellyjac commented:
> On top of that, the license requires that you agree to these use restrictions, or you must remove the software (source and binary) from your hard disks.

https://fedoraproject.org/wiki/Licensing:Unrar

We might actually want to strip the code from the src FOD :eyes:
Maybe that crap doesn't count if it's `.tar.xz`-ed up :shrug:

peterhoeg commented:
From the link:

> Accordingly, this means that any code under this license can not be shipped as part of Fedora in source or binary form.

If you amend the source to fetch from here, it should be safe as I strip the Unrar sources:

https://gitlab.com/peterhoeg/7z-free

06kellyjac committed:
_7zz: correct license and remove p7zip dependency

The code under Compress/Rar* is licensed under a specific unRAR license
Also Compress/LzfseDecoder.cpp is covered by BSD3

The unRAR code is removed from the `.tar.xz` since the license posits you
agree or remove the code from your hard drive
This adds some complexity to updating 7zz so there is also an update
script

Meta has been updated and tweaked

Source is now downloaded from sourceforge in the `.tar.xz` version to
avoid depending on p7zip

06kellyjac committed:
_7zz: add notice of the license updates and optional unRAR licenced code

06kellyjac commented:
I'm not a huge fan of maintaining a repo following 7z without the unRAR code. The nix for grabbing the source and ripping out the unRAR code is a bit more complex but I've added an update script to handle that.

I should also probably sign up to maintain this mess I've made :sweat_smile:

anna328p commented:
should we remove the unRAR code, or have an option in the package to enable it, and expose two packages? i use 7zz to extract RAR files very often.

peterhoeg commented:
There is an option to enable it - it's just off by default.

06kellyjac commented:
exactly, you can either add an overlay to get it with unRAR globally:

```
final: prev: {
  _7zz = prev._7zz.override { enableUnfree = true; };
}
```

or just override it where you add it to your machine:

```
environment.systemPackages = with pkgs; [
  (_7zz.override { enableUnfree = true; })
];
```

06kellyjac commented:
update passthru was pointing at the wrong file. Pushed a fix
